### DISCUSSION TITLE
Is this possible to use installed from vcpkg external libs for processing xar archives?

### DISCUSSION THREAD

**Discussion by kzatsepin** [microsoft/vcpkg#31701] [General]
Hi there! 
I'm trying to process xar files and from archive_read_support_xar.c
I see the comments like this

 * xar needs several external libraries.
 *   o libxml2 or expat --- XML parser
 *   o openssl or MD5/SHA1 hash function
 *   o zlib
 *   o bzlib2 (option)
 *   o liblzma (option)
Also there are [recommendations ](https://github.com/libarchive/libarchive/wiki/BuildInstructions#user-content-Building_on_Windows)to setup needed libs at C:\Program Files\gnuwin32 , but I have needed libs at vcpkg location.
Should I have a patch file of CMakeLists.txt to handle that? And how it has to be done?

**Comment by dg0yt** [comment]
Are you trying to use port libarchive?
Doesn't xar work when you install `libarchive[libxml2,crypto]`?
(Unfortunately, this port's features don't name features but dependencies.)

**Reply by kzatsepin** [reply]
I'm sorry, do you mean to install like
`vcpkg install libarchive[libxml2,crypto]:x64-windows` ?

**Reply by dg0yt** [reply]
Yes.

**Reply by kzatsepin** [reply]
Gonna try it tomorrow, thank you!

**Reply by dg0yt** [reply]
Well, it shouldn't make a difference. Default features turn `libarchive:x64-windows` into 
`libarchive[bzip2,core,crypto,libxml2,lz4,lzma,zstd]:x64-windows`
so `libxml2` and `crypto` are included.

**Reply by kzatsepin** [reply]
So the problem is CMakeLists can not find out these libs, and I have to specify path to these libs, but idk how. I guess I should create a patch file editing CMakeLists.txt in some way

**Reply by kzatsepin** [reply]
Also zlib is needed but that's not in dependencies list

**Reply by dg0yt** [reply]
No, vcpkg brings its own ports to provide libraries.

**Reply by dg0yt** [reply]
> Also zlib is needed but that's not in dependencies list

It *is* in the dependencies. It is *not an optional feature*.
https://github.com/microsoft/vcpkg/blob/1c5a340f6e10985e2d92af174a68dbd15c1fa4e1/ports/libarchive/vcpkg.json#L8-L14

**Reply by kzatsepin** [reply]
Ok, let's imagine it is installed by vcpkg and libarchive can see all of these libs. In the file `archive_read_support_xar.c`
There are lines like:

`#ifdef HAVE_LIBNAME_H`
`#include <libname.h>`
`#endif`

And all of these needed libs are in "hidden" area

And if there are all libs included it makes real function visible, otherwise it is hidden and some stub is used (it returns an error)

As I understand when libarchive are being installed it tries to find path to libs and if it finds it sets
+SET HAVE_LIBNAME_H

I hope I'm right at how it works and you understand me :DD
So why is it hidden then?

**Reply by dg0yt** [reply]
Most of these guard are for optional or alternative dependencies.

The key guard is this:
https://github.com/libarchive/libarchive/blob/ba80276ccc3c941c4918ec6e2460059f0c525c43/libarchive/archive_read_support_format_xar.c#L59-L62

This requires an xml lib (libxml2), zlib and crypto (openssl).

**Comment by kzatsepin** [comment]
I've installed libarchive by 
vcpkg install libarchive:x64-windows
And ofc it installs all of the dependencies it needs by its own. But as I understand when libarchive installs it tries to use some libs from C/Program Files(x86)/gnuwin32 cuz I saw some lines at CmakeLists.txt that tries to get path to libs need to process xar archives from this path. 
And the problem is that some libs to process xar are not in dependency list of libarchive. If these libs are not found, it uses the second variant of function which is throwing an error like "xar processing is not possible on this platform"

**Reply by dg0yt** [reply]
> But as I understand when libarchive installs it tries to use some libs from C/Program Files(x86)/gnuwin32 cuz I saw some lines at CmakeLists.txt that tries to get path to libs need to process xar archives from this path. And the problem is that some libs to process xar are not in dependency list of libarchive. If these libs are not found, it uses the second variant of function which is throwing an error like "xar processing is not possible on this platform"

The gnuwin32 installation is one general way to provide these libraries, but it isn't needed or meant to be used in vcpkg.
Of course something could go wrong. This sould be visible in the config logs or CMakeCache.
